modules/renter/contractor/uptime.go

```diff
@@ -0,0 +1,76 @@
+package contractor
+
+import (
+	"time"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/modules"
+)
+
+// uptimeMinScans is the minimum number of scans required to judge whether a
+// host is offline or not.
+const uptimeMinScans = 3
+
+// uptimeWindow specifies the duration in which host uptime is checked.
+var uptimeWindow = func() time.Duration {
+	switch build.Release {
+	case "dev":
+		return 30 * time.Minute
+	case "standard":
+		return 7 * 24 * time.Hour // 1 week
+	case "testing":
+		return 15 * time.Second
+	}
+	panic("undefined uptimeWindow")
+}()
+
+// isOffline decides whether a host should be considered offline, based on its
+// scan metrics.
+func isOffline(host modules.HostDBEntry) bool {
+	// consider a host offline if:
+	// 1) The host has been scanned at least three times, and
+	// 2) The three most recent scans have all failed, and
+	// 3) The time between the most recent scan and the last successful scan
+	//    (or first scan) is at least uptimeWindow
+	numScans := len(host.ScanHistory)
+	if numScans < uptimeMinScans {
+		// not enough data to make a fair judgment
+		return false
+	}
+	// NOTE: ScanHistory is ordered from oldest-newest
```

Comment:
ok, consider this case:

- user scans host three times within an hour. All three scans fail, but the host is not marked offline because the scans did not happen over a sufficient window.
- user sets their clock back 1 week
- user scans again, this time successfully
- scan is inserted in sorted order, so it appears 1 week before the first three scans
- now we have three recent failed scans, and the last successful scan was over a week ago. The host is marked offline.

So now a host can be marked offline even if the user _just scanned it successfully!_ Clearly this behavior is not correct either.

It's misleading to say that this code is "relying" on the user not adjusting their clock, as though most of our code did not make such assumptions. We implicitly rely on the user not to do hundreds of things. It's not our responsibility to anticipate and guard against every way the user might modify their system. We can only focus on the common cases. For example, it's not uncommon for users to hard-power-off their systems, so we guard against that by using atomic files, transactional databases, etc. But it is very uncommon for a user to set their `ulimit` to a number small enough to prevent siad from opening its persist files. When that happens, all we can say is "well, it's not reasonable to expect siad to work if you do that." Likewise, it's not reasonable to expect a distributed consensus application to function correctly if your clock is set a week in the past.